providers/asrockrack/firmware.go

```diff
@@ -0,0 +1,221 @@
+package asrockrack
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/bmc-toolbox/bmclib/devices"
+	bmclibErrs "github.com/bmc-toolbox/bmclib/errors"
+)
+
+// FirmwareInstall uploads and initiates firmware update for the component
+func (a *ASRockRack) FirmwareInstall(ctx context.Context, component, applyAt string, forceInstall bool, reader io.Reader) (jobID string, err error) {
+	var size int64
+	if file, ok := reader.(*os.File); ok {
+		finfo, err := file.Stat()
+		if err != nil {
+			a.log.V(2).Info("warn", "unable to determine file size: "+err.Error())
+		}
+
+		size = finfo.Size()
+	}
+
+	switch component {
+	case devices.SlugBIOS:
+		err = a.firmwareInstallBIOS(ctx, reader, size)
+	case devices.SlugBMC:
+		err = a.firmwareInstallBMC(ctx, reader, size)
+	default:
+		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstall, "component unsupported: "+component)
+
+	}
+
+	return jobID, err
+}
+
+// FirmwareInstallStatus returns the status of the firmware install process, a bool value indicating if the component requires a reset
+func (a *ASRockRack) FirmwareInstallStatus(ctx context.Context, component, installVersion, taskID string) (status string, err error) {
+	switch component {
+	case devices.SlugBIOS:
+		return a.firmwareUpdateBIOSStatus(ctx, installVersion)
+	case devices.SlugBMC:
+		return a.firmwareUpdateBMCStatus(ctx, installVersion)
+	default:
+		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstall, "component unsupported: "+component)
+	}
+}
+
+// firmwareInstallBMC uploads and installs firmware for the BMC component
+func (a *ASRockRack) firmwareInstallBMC(ctx context.Context, reader io.Reader, fileSize int64) error {
+	var err error
+
+	// 1. set the device to flash mode - prepares the flash
+	a.log.V(2).Info("info", "action", "set device to flash mode, takes a minute...", "step", "1/4")
+	err = a.setFlashMode(ctx)
+	if err != nil {
+		return fmt.Errorf("failed in step 1/4 - set device to flash mode: " + err.Error())
```

Comment:
Why not use `errors.Wrap()`? That would allow callers to use `errors.Is()`. Same for the other steps in this file

Reply:
Thanks for the suggestion here, as of now the file size is required for the multipart upload. I've created a separate issue to follow up on the LengthReader interface https://github.com/bmc-toolbox/bmclib/pull/261#discussion_r856265780

Fixed up errors Wrap in [aa25390](https://github.com/bmc-toolbox/bmclib/pull/267/commits/aa2539090556e1566231293846e85651fd4d682b)